design-proposals/klip-3-serialization-of-single-fields.md

```diff
@@ -0,0 +1,1076 @@
+including records and top-level primitives, arrays and maps.
+
+The format requires KSQL to be configured to store and retrieve the Avro schemas from the |sr-long|.
+For more information, see :ref:`install_ksql-avro-schema`.
+
+------------
+Avro Records
+------------
+
+Avro records can be deserialized into matching KSQL schemas.
+
+For example, given a KSQL statement such as:
+
+.. code:: sql
+
+    CREATE STREAM x (ID BIGINT, NAME STRING, AGE INT) WITH (VALUE_FORMAT='JSON', ...);
+
+And an Avro record serialized with the schema:
+
+.. code:: json
+
+       {
+         "type": "record",
+         "namespace": "com.acme",
+         "name": "UserDetails",
+         "fields": [
+           { "name": "id", "type": "long" },
+           { "name": "name", "type": "string" }
+           { "name": "age", "type": "int" }
+         ]
+       }
+
+KSQL will deserialize the Avro record's fields into the corresponding fields of the stream.
+
+.. note:: Where the key or value schema contains only a single field KSQL, by default, 
+          will both expect source data and produce sink data where the field has been 
+          serialized as an anonymous value, rather than as a named field within an 
+          Avro record.
+          
+          For more information, see :ref:`single_field_wrapping`.
+
+-------------------------------------
+Top-level primitives, arrays and maps
+-------------------------------------
+
+The Avro format supports reading top-level primitives, arrays and maps, but only if the target
+schema contains a single field of a compatible type.
+
+For example, given a KSQL statement with only a single field in the value schema:
+
+.. code:: sql
+
+    CREATE STREAM x (ID BIGINT) WITH (VALUE_FORMAT='JSON', ...);
+
+And an Avro value serialized with the schema:
+
+.. code:: json
+
+       {
+         { "type": "long" }
+       }
+
+KSQL will deserialize the values into the ``ID`` field of the stream.
+
+When serializing data with a single fields, KSQL will also serialize the
+field as an anonymous value, unless the default behaviour has been overridden. 
+For more information, see :ref:`single_field_wrapping`.
+
+=========================
+Controlling serialization
+=========================
+
+KSQL offers several mechanisms for controlling serialization and deserialization.
+
+The primary mechanism is by choosing the serialization format when you create
+a stream or table and specify the ``VALUE_FORMAT`` in the ``WITH`` clause.
+
+.. code:: sql
+
+    CREATE TABLE x (F0 INT, F1 STRING) WITH (VALUE_FORMAT='JSON', ...);
+
+For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
+
+KSQL provides some additional configuration that allows serialization to be controlled:
+
+.. _single_field_wrapping
+-------------------------
+Single field (un)wrapping
+-------------------------
+
+.. note:: The ``DELIMITED`` format is not effected by single field unwrapping.
```

Comment:
Out of curiosity, why do we prefer to throw an error if a user tries to specify `WRAP_SINGLE_XXX` on a statement with multiple key/value fields, but not if a user tries to specify `WRAP_SINGLE_XXX` on a `DELIMITED` topic?

Reply:
Good point... I'll throw for both.